tmt/steps/provision/testcloud.py

```diff
@@ -9,6 +9,7 @@
 import threading
 import types
 from collections.abc import Iterator
+from string import Template
```

Comment:
If dropping `format()` in favor of "templates", how about using Jinja? It's the "lingua franca" when it comes to rendering text blobs in tmt, it's used for docs and test/plan/story templates and paths and export, and using it would keep the number of involved "templating solutions" on the same number it is now.

ha said, I'm not sure whether `safe_substitute()` is important for your use case. Jinja would replace an unknown variable with an empty string rather than preserving it, no idea whether you need to run the template through several rounds of evaluation.

Reply:
Yeah, the per partes replacement is used right now in the code paths (first, the tmt replaces bits like ssh key, then testcloud replaces the rest in its instance.prepare). 

Sigh... it's a good point, I am thinking about something in lines of having this as an interrim solution to allow other work to land, and then continue with replacing/adding/amending api so it can be converted to jinja properly. idk how you'd feel about that?

Reply:
> Yeah, the per partes replacement is used right now in the code paths (first, the tmt replaces bits like ssh key, then testcloud replaces the rest in its instance.prepare()).

Is there some custom/default relationship between these two? Like `testcloud` plugin ignoring keys that `testcloud` package would then replace if they remain available?

I believe Jinja could support this too, there are two options:

* using `default` filter, `{{ runcommands | default("{{ runcommands }}") }}`. When undefined, it would emit `{{ runcomands }}` for the next round of rendering. Ugly, but might work.
* `DebugUndefined`

  ```python
  env = Environment(undefined=DebugUndefined)
  "{{ a }} is defined, but {{ b}} is undefined".render({"a":"a"})

  # produces:
  "a is defined, but {{ b }} is undefined"
  ```
  We would have to fix `tmt.utils.render_template()` to support this "mode", but doable.

> Sigh... it's a good point, I am thinking about something in lines of having this as an interrim solution to allow other work to land, and then continue with replacing/adding/amending api so it can be converted to jinja properly. idk how you'd feel about that?

Not perfect, but I do it too. As long as I have a good idea how the eventual result should look like, it makes sense to produce less-than-ideal intermediate steps to unblock the overall progress. With comments, describing why things are what they are, mentioning the roadmap ("yep, will fix it in the next decade"), in my experience it's a viable approach.

Anyway. If it's a part of a longer project, it makes sense to wrap up things later, although Jinja may still be usable. On the other hand, Jinja is tmt's pick, not testcloud's - if testcloud package developers decide `string.Template` is the best option, `testcloud` plugin developers would need to work with it :)

Reply:
Yeah, after some experimenting, I'd prefer to jump to string.Template as an interim solution, I've added a TODO around the respective area.